consensus/hotstuff/verification/common.go

```diff
@@ -66,3 +68,77 @@ func checkVotesValidity(votes []*model.Vote) error {
 
 	return nil
 }
+
+// stakingKeysAggregator is a structure that aggregates the staking
+// public keys for QC verifications.
+type stakingKeysAggregator struct {
+	lastStakingSigners map[*flow.Identity]struct{}
+	lastStakingKey     crypto.PublicKey
+	lock               sync.Mutex
+}
+
+// creates a new staking keys aggregator
+func newStakingKeysAggregator() *stakingKeysAggregator {
+	aggregator := &stakingKeysAggregator{
+		lastStakingSigners: map[*flow.Identity]struct{}{},
+		lastStakingKey:     crypto.NeutralBLSPublicKey(),
+		lock:               sync.Mutex{},
+	}
+	return aggregator
+}
+
+// aggregatedStakingKey returns the aggregated public of the input signers.
+func (s *stakingKeysAggregator) aggregatedStakingKey(signers flow.IdentityList) (crypto.PublicKey, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	// this greedy algorithm assumes the signers set does not vary much from one call
+	// to aggregatedStakingKey to another. It computes the delta of signers compared to the
+	// latest list of signers and adjust the latest aggregated public key. This is faster
+	// than aggregating the public keys from scratch at each call.
+	//
+
+	// get the signers delta and update the last list for the next comparison
+	newSigners, missingSigners := s.identitiesDelta(signers)
+	// add the new keys
+	var err error
+	s.lastStakingKey, err = crypto.AggregateBLSPublicKeys(
+		append(newSigners.StakingKeys(), s.lastStakingKey))
+	if err != nil {
+		return nil, fmt.Errorf("adding new staking keys failed: %w", err)
+	}
+	// remove the missing keys
+	s.lastStakingKey, err = crypto.RemoveBLSPublicKeys(s.lastStakingKey, missingSigners.StakingKeys())
+	if err != nil {
+		return nil, fmt.Errorf("removing missing staking keys failed: %w", err)
+	}
+	return s.lastStakingKey, nil
+}
+
+// identitiesDelta computes the delta between the reference s.lastStakingKey
+// and the input identity list
+func (s *stakingKeysAggregator) identitiesDelta(signers flow.IdentityList) (flow.IdentityList, flow.IdentityList) {
+	var newSigners, missingSigners flow.IdentityList
+
+	// create a map of the input list,
+	// and check the new signers
+	signersMap := map[*flow.Identity]struct{}{}
```

Comment:
Yep, good idea!

Reply:
Isn't it a lighter to use the pointer instead of the entire struct? Do you think the pointer may not be unique (the same identity represented by 2 objects and hence 2 pointers) ?

Reply:
I don't think we should be relying on the pointers being identical. We generally don't enforce the singleton pattern for Identities in the code. Meaning, we might have two different instances for the same `identity`.

Reply:
Yeah, pointer is lighter but not too coupled with the implementation. I still think using flow.Identifier as the key is better